tuf/client/client.go

```diff
@@ -45,44 +45,77 @@ func (c *Client) Update() error {
 	// 3. Check if root correct against snapshot
 	//   a. If incorrect, download new root and return to 1.
 	// 4. Iteratively download and search targets and delegations to find target meta
+
+	type TufUpdateState int
+
+	const (
+		StateDownloadTimestamps TufUpdateState = iota
+		StateDownloadSnapshots  TufUpdateState = iota
+		StateDownloadTargets    TufUpdateState = iota
+		StateCheckRoot          TufUpdateState = iota
+		StateDownloadRoot       TufUpdateState = iota
+		StateUpdateComplete     TufUpdateState = iota
+		StateUpdateFailed       TufUpdateState = iota
+	)
+
+	var err error
+	downloadedRoot := false
+	downloadState := StateDownloadTimestamps
 	logrus.Debug("updating TUF client")
-	err := c.update()
-	if err != nil {
-		logrus.Debug("Error occurred. Root will be downloaded and another update attempted")
-		if err := c.downloadRoot(); err != nil {
-			logrus.Debug("Client Update (Root):", err)
-			return err
-		}
-		// If we error again, we now have the latest root and just want to fail
-		// out as there's no expectation the problem can be resolved automatically
-		logrus.Debug("retrying TUF client update")
-		return c.update()
-	}
-	return nil
-}
 
-func (c *Client) update() error {
-	err := c.downloadTimestamp()
-	if err != nil {
-		logrus.Debugf("Client Update (Timestamp): %s", err.Error())
-		return err
-	}
-	err = c.downloadSnapshot()
-	if err != nil {
-		logrus.Debugf("Client Update (Snapshot): %s", err.Error())
-		return err
-	}
-	err = c.checkRoot()
-	if err != nil {
-		// In this instance the root has not expired base on time, but is
-		// expired based on the snapshot dictating a new root has been produced.
-		logrus.Debug(err)
-		return err
+	for downloadState != StateUpdateComplete && downloadState != StateUpdateFailed {
+		switch downloadState {
+		case StateDownloadTimestamps:
+			err = c.downloadTimestamp()
+			if err != nil {
+				logrus.Debugf("Client Update (Timestamp): %s", err.Error())
+				downloadState = StateDownloadRoot
+				break
```

Comment:
No need for the `break` statements. Go switches explicitly do not fall through. Instead of adding `break` to exit the switch, it's the default behaviour and you have to add `fallthrough` to force it to fall through.